src/simulator_core/entities/assets/demand_cluster.py

```diff
@@ -29,50 +54,159 @@ def __init__(
+        self.temperature_return_target = setpoints[PROPERTY_TEMPERATURE_RETURN]
+
+        if self.simulation_performed():
+            self.temperature_supply = self.pandapipes_net.res_flow_control['t_from_k'][
+                self._flow_control.index]
+            self.temperature_return = self.pandapipes_net.res_heat_exchanger['t_to_k'][
+                self._heat_exchanger.index]
+            self.mass_flowrate = self.pandapipes_net.res_heat_exchanger['mdot_from_kg_per_s'][
+                self._heat_exchanger.index]
+
+        # adjust flowrate or power to meet the return temperature
+        if self.pandapipes_net.flow_control.control_active[self._flow_control.index] is True:
+            # if pump is active, set flowrate to meet Target Temperature
+            self._heat_exchanger.qext_w = self.thermal_power_allocation
+            self.pandapipes_net.heat_exchanger.qext_w[
+                self._heat_exchanger.index] = self.thermal_power_allocation
+
+            adjusted_mass_flowrate = heat_demand_and_temperature_to_mass_flow(
+                self.thermal_power_allocation,
+                self.temperature_supply,
+                self.temperature_return_target,
+                self.pandapipes_net)
+
+            self._flow_control.controlled_mdot_kg_per_s = adjusted_mass_flowrate
+            self.pandapipes_net.flow_control.controlled_mdot_kg_per_s[
+                self._flow_control.index] = adjusted_mass_flowrate
+
+        else:
+            # if pump is disabled, set thermal power to meet Target Temperature
+            adjusted_thermal_power_demand = mass_flow_and_temperature_to_heat_demand(
+                self.temperature_supply,
+                self.temperature_return_target,
+                self.mass_flowrate,
+                self.pandapipes_net)
+
+            self._heat_exchanger.qext_w = adjusted_thermal_power_demand
+            self.pandapipes_net.heat_exchanger.qext_w[
+                self._heat_exchanger.index] = adjusted_thermal_power_demand
+
+    def get_setpoints(self) -> Dict:
         """Placeholder to get the setpoint attributes of an asset.
 
         :return Dict: The setpoints of the asset. The keys of the dictionary are the names of the
             setpoints and the values are the values.
         """
-        return {}
+        setpoints = dict()
+        setpoints[PROPERTY_HEAT_DEMAND] = self.thermal_power_allocation
+        setpoints[PROPERTY_TEMPERATURE_RETURN] = self.temperature_return
+
+        return setpoints
 
     def simulation_performed(self) -> bool:
-        """Placeholder to indicate that a simulation has been performed.
+        """Check if the simulation has been performed.
 
-        :return bool: True if a simulation has been performed, False otherwise.
+        :return bool simulation_performed: True if the simulation has been performed,
+            False otherwise.
         """
-        return True
+        return self._simulated
 
```

Comment:
Can you make an issue for this method? I sort of like it but if you implement it here you should also change all the other assets. For now keep it similar to the initial assets? 

What do you guys think @samvanderzwan @MichielTukker?

Reply:
Make an issue so we can refactor this later